doc/dev/code-generator.md

```diff
@@ -0,0 +1,211 @@
+
+In `example.cpp` and `example.hpp` we create our subclass of `GenTemplate`. Therefore `example.hpp` should look like this:
+
+```cpp
+#ifndef ELEKTRA_EXAMPLE_HPP // choose a unique header guard
+#define ELEKTRA_EXAMPLE_HPP
+
+#include "../template.hpp"
+
+class ExampleGenTemplate : public GenTemplate
+{
+public:
+    ExampleGenTemplate () : GenTemplate ("example", { ".txt" }, {}, {})
+    {
+    }
+
+protected:
+    kainjow::mustache::data getTemplateData (const std::string & outputName, const std::string & part, const kdb::KeySet & ks,
+                                             const std::string & parentKey) const override;
+};
+
+#endif // ELEKTRA_EXAMPLE_HPP
+```
+
+Apart from the line `ExampleGenTemplate () : GenTemplate ("example", { ".txt" }, {}, {})` everything should be more or less the same for
+all templates. Let's dissect this line:
+
+- `ExampleGenTemplate ()` we declare a zero argument constructor. All templates _must_ have only a single zero argument constructor,
+  otherwise the framework cannot instantiate them.
+- `: GenTemplate (` the constructor must invoke the base-class constructor.
+- `"example",` we decided to use `example` as the base name for our template files. This base name will be replaced by the `outputName`
+  chosen by the user, when invoking the code-generator.
+- `{ ".txt" },` this is the list of part suffixes for our template. We only have a single part (output file) with the suffix `.txt`.
+- `{},` the first empty list would contain all the partials our template uses. We don't use any.
+- `{},` the second empty list, is actually a map. It would contain all parameters and whether they are required or not.
+  We don't have parameters.
+
+As you can see, the header is quite simple. The more important part is actually the source file. It contains the implementation of
+`getTemplateData`, the function that delivers all the template data to the framework.
+
+For our example we will use this implementation:
+
+```cpp
+kainjow::mustache::data ExampleGenTemplate::getTemplateData (const std::string & outputName, const std::string & part,
+                                                             const kdb::KeySet & ks, const std::string & parentKey) const
+{
+    using namespace kdb;
+    using namespace kainjow::mustache;
+
+    list keyList;
+
+    for (auto it = ks.begin (); it != ks.end (); ++it)
+    {
+        Key key = *it;
+
+        auto keyObject = object({ { "name", key.getName() } });
+
+        keyList.emplace_back(keyObject);
+    }
+
+    auto data = object({ { "keys", keyList } });
+
+    return data;
+}
+```
+
+The framework will invoke `getTemplateData` for each part of our template with the `outputName` and `parentKey` as given on the command line,
+as well as the current part suffix (`part`) and the input keyset `ks`. All keys of the input keyset are guaranteed to be below `parentKey`.
+
+How exactly the above code works is left as an exercise to the reader.
```

Comment:
Explain or keep simpler.

Reply:
It won't get much simpler than iterating through a `KeySet` simply collecting the key names.